pkg/util/merger.go

```diff
@@ -0,0 +1,50 @@
+package util
+
+import "github.com/prometheus/common/model"
+
+// MergeSampleSets merges and dedupes two sets of already sorted sample pairs.
+func MergeSampleSets(a, b []model.SamplePair) []model.SamplePair {
+	result := make([]model.SamplePair, 0, len(a)+len(b))
+	i, j := 0, 0
+	for i < len(a) && j < len(b) {
+		if a[i].Timestamp < b[j].Timestamp {
+			result = append(result, a[i])
+			i++
+		} else if a[i].Timestamp > b[j].Timestamp {
+			result = append(result, b[j])
+			j++
+		} else {
+			result = append(result, a[i])
+			i++
+			j++
+		}
+	}
+	for ; i < len(a); i++ {
+		result = append(result, a[i])
+	}
+	for ; j < len(b); j++ {
+		result = append(result, b[j])
+	}
+	return result
+}
+
+// MergeNSampleSets merges and dedupes n sets of already sorted sample pairs.
+func MergeNSampleSets(sampleSets ...[]model.SamplePair) []model.SamplePair {
+	l := len(sampleSets)
+	switch l {
+	case 0:
+		return []model.SamplePair{}
+	case 1:
+		return sampleSets[0]
+	}
+
+	c := make(chan []model.SamplePair, l)
+	for _, ss := range sampleSets {
+		c <- ss
+	}
+	for ; l > 1; l-- {
+		left, right := <-c, <-c
+		c <- MergeSampleSets(left, right)
+	}
+	return <-c
```

Comment:
Use of the channel seems a bit gratuitous here.  Wouldn't a slice work just as well?  Or even the recursion approach? I guess its not a big deal.